### ISSUE TITLE
Tiling using Pytorch Higher-Order Operators

### ISSUE THREAD

**Issue by tardieu** [torch-spyre/torch-spyre#3965]
PyTorch has recently added support for higher-order operators such as `while_loop` and `map`. Higher-order operators should permit a PyTorch-native approach to tiling (at the expense of depending on a still-evolving API).

`while_loop` makes it possible to implement loops that are preserved in the FX Graph and during lowering, producing a nested graph structure with a subgraph implementing the loop body.

Using a `while_loop` it is possible to iterate over the rows of a tensor or a number of tiling dimensions. The `while_loop` operator however does not support tiling out of the box as using iteration variables to slice a tensor is not supported. Constructing mask tensors using these iteration variables is possible but defeats the purpose of computing over tiles.

The `map` operator applies a function to each tensor row. Its implementation demonstrates how to work around the `while_loop` limitations by:
- enforcing a rigid tile structure (tile == row). The loop body has no control over the slicing.
- tracing the loop body for the 1st tile only (i.e., using the 1st row as an example input),
- constructing a while loop IR directly,
- “cheating” by adding increasing offsets to subgraph input/output tensors for each iteration.

In order to permit tiling in torch-spyre by means of higher-order operators we need to add support for:
1. compiling subgraphs (ie compiling multiple graphs and handling parent/child relationships),
2. handling WhileLoop IR nodes (and implicit back edges),
3. generalizing `map` to multiple tensors, multiple dimensions, and non-unit strides.

Concretely we want to introduce a `for_each_tile` op operating on tiles.

As a first step this `for_each_tile` op can explicitly specify the affine offset expressions for each input/output of the subgraph. Later this can be combined with our existing `propagate_named_dims` to automatically derives these expressions from the conceptual tiling dimensions and tile counts. A significant chunk of the existing working set reduction implementation (WSR) should be immediately applicable to this.

In contrast to WSR, loop bodies will be written in terms of tiles rather than tensors, specifically as functions operating on arguments bound to tiles at runtime.

The `map` operator does not guarantee the iteration order making it possible to support parallel execution. While a `for_each_tile` operator could follow and be amenable to both core division and working set reduction, this is not something we will consider initially. We will instead focus on sequential use cases such as online softmax. The `for_each_tile` operator unlike `map` and like `while_loop` will have explicitly "carried tensors" making it possible to not only accumulate across iterations but also have access to the already accumulated values (the partial result) from within the loop body.

In contrast with WSR, this approach does not implicitly handle split reductions. They have to be programmed explicitly using temporary tensors and explicit add/max/… ops. This automatic treatment of split reduction is something we can reconsider at a later date.

In WSR the lowered graph is still a flat op sequence. The loop structure is carried via custom metadata and opaque to PyTorch leading to issues with ordering optimizations. They necessitate compensation mechanisms such as the `reorder_unhinted_interlopers` pass. A native loop representation in PyTorch will eliminate this gap.

The torch-spyre OpSpec representation of a WhileLoop should be very similar to the existing LoopSpec and major codegen changes (SuperDSC or KTIR) are not expected.

Many custom compiler passes in torch-spyre are already built with the idea that a single iteration of a loop is enough to represent the entirety of the compilation task (memory allocation, work division across cores, etc). Since PyTorch higher-order operators and subgraphs adopt the same approach, they should be easy to port over, with the only added complexity of handling carried tensors (essentially tile-sized tensors scoped to the whole loop instead of a single iteration).

**Comment by tardieu** [comment]
Risk analysis:
One challenge with HOPs is the dependency on fast evolving experimental code. The `scan` and `associative_scan` implementations as HOPs are a point in favor. The existing `map` implementation, while experimental, suggests that partitioning a tensor to iterate over is a valid use case. A `for_each_tile` construct has the potential to be upstreamed like `scan`. 

Hybrid alternatives:
We could consider hybrid alternatives for instance combining a more mature HOP like `while_loop` with spyre hints, where hints would be used to define the tile strides without having to resort to slicing arithmetic in the loop body. Moreover, we could target the IR directly (`WhileLoop`) without relying on HOPs/custom HOPs at the source level. Again, this would require some form of control flow construct akin to spyre hints.

**Comment by tardieu** [comment]
Summary of discussion with @bohnstingl:

We should be able to prototype `for_each_tile` by wrapping `scan` and pre/postfixing the `scan` invocation with permutations of the input and output tensors. We can start with such a prototype and then move to a direct implementation to avoid the data copies.

The backend compiler for Spyre today requires loop counts to be decided before the kernel invocation, hence cannot support arbitrary while loops. This is not an issue for a tiling use case since the tile counts are known at invocation time (even if symbolic during compilation). We may want to maintain the loop count expression as custom metadata on the `WhileLoop` IR and ignore the termination condition subgraph during codegen. In general, we may try to recognize and support `while_loop` instances with pre-computable loop counts, but this is out of scope for now.

Support for indirectly accessed tensors is required but should not require additional logic. An index tensor can be tiled like any other tensor. An indirectly accessed value tensor is never tiled.

**Comment by tardieu** [comment]
Summary of discussion with @dgrove-oss and @marnold-ibm:

A `WhileLoop` includes a subgraph for the termination condition and a subgraph for the loop body. The loop condition and loop body are functional: the tensor arguments and return values are never aliased. The implementation of the loop back edge is the responsibility of the inductor backend. For example the cpp backend implements the `WhileLoop` as a cpp `while` loop and feeds the return values as arguments to the next iteration using pointer assignments. On Spyre we will need to rely on tensor to tensor copies (`identity` superDSCs) and rely on the memory allocator or a compiler pass to eliminate these copies when possible.

**Comment by bohnstingl** [comment]
@tardieu, I prepared an initial sketch for a `for_each_tile` op that can be used to represent tiled loops. Happy to hear your thoughts on it.

**Comment by dgrove-oss** [comment]
Closing as completed.
